### PR TITLE
Optimize checking incremental traits by using NSSet

### PR DESCRIPTION
**Pull request by ladanazita**

### CODE REVIEW

File: Pod/Classes/SEGAmplitudeIntegration.m [hunk 1]
```diff
@@ -17,6 +17,7 @@ - (id)initWithSettings:(NSDictionary *)settings andAmplitude:(Amplitude *)amplit
         self.amplitude = amplitude;
         self.amprevenue = amprevenue;
         self.identify = identify;
+        self.traitsToIncrement = [NSSet setWithArray:self.settings[@"traitsToIncrement"]];
```
Comment (f2prateek) [inline Pod/Classes/SEGAmplitudeIntegration.m]: what if self.settings is nil/undefined?

File: Pod/Classes/SEGAmplitudeIntegration.m [hunk 2]
```diff
@@ -199,22 +199,18 @@ - (void)reset
...
+        } else {
+            [self.amplitude identify:[self.identify set:trait value:value]];
+            SEGLog(@"[Amplitude set:%@ value:%@]", trait, value);
         }
-    }
-
-    if (!isAmountSet) {
-        [self.amplitude identify:[self.identify set:trait value:value]];
-    }
+    };
```
Comment (f2prateek) [inline Pod/Classes/SEGAmplitudeIntegration.m]: extra `;`?

File: Pod/Classes/SEGAmplitudeIntegration.m [hunk 3]
```diff
@@ -199,22 +199,18 @@ - (void)reset
...
-- (void)incrementTrait:(NSString *)trait andValue:(NSString *)value
+- (void)incrementOrSetTraits:(NSDictionary *)traits
 {
-    __block BOOL isAmountSet = false;
-
-    NSArray *increments = self.settings[@"traitsToIncrement"];
-    for (NSString *increment in increments) {
-        if ([increment isEqualToString:trait]) {
+    for (NSString *trait in traits) {
+        id value = [traits valueForKey:trait];
```
Comment (f2prateek) [inline Pod/Classes/SEGAmplitudeIntegration.m:207]: Is it ok to use `id`?

Comment (ladanazita) [reply Pod/Classes/SEGAmplitudeIntegration.m:207]: Not sure what the value will be for each `trait`

Comment (f2prateek) [reply Pod/Classes/SEGAmplitudeIntegration.m:207]: Yes, but is it ok to call `set`/`add` on Amplitude and pass it an `id`?

File: Pod/Classes/SEGAmplitudeIntegration.m (hunk 1 above)
Comment (f2prateek) [reply Pod/Classes/SEGAmplitudeIntegration.m]: (I think it should be ok, but worth confirming the behaviour).

**Review by f2prateek — APPROVED** [review]
Some style changes. Would love to confirm the behaviour for these 2 things:

1. if the traitsToIncrement setting is undefined.
2.  if it's ok to call `set`/`add` with `id` as the type - particularly if the object. In the android one, Amplitude only accepts certain types (strings, doubles etc).

**Comment by ladanazita** [comment]
For `add`, they accept only type NSString or NSNumber and handle this check themselves:

```- (AMPIdentify*)add:(NSString*) property value:(NSObject*) value
{
    if ([value isKindOfClass:[NSNumber class]] || [value isKindOfClass:[NSString class]]) {
        [self addToUserProperties:AMP_OP_ADD property:property value:value];
    } else {
        AMPLITUDE_LOG(@"Unsupported value type for ADD operation, expecting NSNumber or NSString");
    }
    return self;
}
```

`set` is a bit more complex, and eventually gets to a check for each type and executes additional logic given the type.

File: Pod/Classes/SEGAmplitudeIntegration.m (hunk 3 above)
Comment (ladanazita) [reply Pod/Classes/SEGAmplitudeIntegration.m:207]: For `add`, they accept only type NSString or NSNumber and handle this check themselves:

```- (AMPIdentify*)add:(NSString*) property value:(NSObject*) value
{
    if ([value isKindOfClass:[NSNumber class]] || [value isKindOfClass:[NSString class]]) {
        [self addToUserProperties:AMP_OP_ADD property:property value:value];
    } else {
        AMPLITUDE_LOG(@"Unsupported value type for ADD operation, expecting NSNumber or NSString");
    }
    return self;
}
```

`set` is a bit more complex, and eventually gets to a check for each type and executes additional logic given the type.

Comment (f2prateek) [reply Pod/Classes/SEGAmplitudeIntegration.m:207]: So to confirm, we can call them with any type, and internally they handle it properly or bail gracefully if they can't?

If that's the case then we don't need to anything.

Comment (ladanazita) [reply Pod/Classes/SEGAmplitudeIntegration.m:207]: Yep

**Pull request merged**